From a4b357dfb4ecae2c40189d1e6f9f992dce58f782 Mon Sep 17 00:00:00 2001
From: Naveen Mamindlapalli <naveenm@marvell.com>
Date: Fri, 23 Jul 2021 14:28:05 +0530
Subject: [PATCH 1630/1921] octeontx2-pf: Fix incorrect MCAM entry index

Fix incorrect MCAM entry index calculation when adding a new
tc flow entry.

Fixes: 4f463d2e38f5 ("octeontx2-pf: Add tc flower hardware offload on ingress traffic")
Change-Id: I3f0cf8d776cd8f4c86b6eba1ff2c0612e9922ca4
Signed-off-by: Naveen Mamindlapalli <naveenm@marvell.com>
Reviewed-on: https://sj1git1.cavium.com/c/IP/SW/kernel/linux/+/57095
Reviewed-by: Sunil Kovvuri Goutham <sgoutham@marvell.com>
Tested-by: Sunil Kovvuri Goutham <sgoutham@marvell.com>
[WK: The original patch got from Marvell sdk11.21.09]
Signed-off-by: Wenlin Kang <wenlin.kang@windriver.com>
---
 drivers/net/ethernet/marvell/octeontx2/nic/otx2_flows.c | 1 +
 drivers/net/ethernet/marvell/octeontx2/nic/otx2_tc.c    | 2 +-
 2 files changed, 2 insertions(+), 1 deletion(-)

diff --git a/drivers/net/ethernet/marvell/octeontx2/nic/otx2_flows.c b/drivers/net/ethernet/marvell/octeontx2/nic/otx2_flows.c
index 70c5cd090618..cd12618ecf66 100644
--- a/drivers/net/ethernet/marvell/octeontx2/nic/otx2_flows.c
+++ b/drivers/net/ethernet/marvell/octeontx2/nic/otx2_flows.c
@@ -125,6 +125,7 @@ static int otx2_alloc_ntuple_mcam_entries(struct otx2_nic *pfvf, u16 count)
 
 	flow_cfg->ntuple_offset = 0;
 	flow_cfg->ntuple_max_flows = allocated;
+	flow_cfg->tc_flower_offset = flow_cfg->ntuple_offset;
 	flow_cfg->tc_max_flows = allocated;
 
 	if (allocated != count)
diff --git a/drivers/net/ethernet/marvell/octeontx2/nic/otx2_tc.c b/drivers/net/ethernet/marvell/octeontx2/nic/otx2_tc.c
index 6545fcd8cf08..4a2bd89c7477 100644
--- a/drivers/net/ethernet/marvell/octeontx2/nic/otx2_tc.c
+++ b/drivers/net/ethernet/marvell/octeontx2/nic/otx2_tc.c
@@ -693,7 +693,7 @@ static int otx2_tc_add_flow(struct otx2_nic *nic,
 					       nic->flow_cfg->tc_max_flows);
 	req->channel = nic->hw.rx_chan_base;
 	req->entry = nic->flow_cfg->flow_ent[nic->flow_cfg->tc_flower_offset +
-				nic->flow_cfg->tc_max_flows - new_node->bitpos];
+				nic->flow_cfg->tc_max_flows - new_node->bitpos - 1];
 	req->intf = NIX_INTF_RX;
 	req->set_cntr = 1;
 	new_node->entry = req->entry;
-- 
2.31.1

